[숫자 야구 게임] 김영우 미션 제출합니다

seank200 opened the pull request:
1주차 미션 제출합니다

  - Java 개발이 매우 오랜만입니다..🥲 "Java스럽지 못한" 부분에 대한 많은 지적과 리뷰를 환영합니다
  - 유저의 답을 처리하는 로직을 `Game`과 `Hint` 클래스로 분리하였습니다.
  - IntelliJ에서의 "Local variable reassigned" 경고를 잘 이해하지 못하였습니다. Java에서는 변수 값을 재지정하는 경우를 최대한 피해야 하는 건가요?

Commit:
feat/implement Guess

Commit:
feat/implement Main app

Commit:
refactor/general maintenance

- Move continue logic to Application.askContinue()
- Remove while(true)

Commit:
chore/add comments

Commit:
chore/add comments

Commit:
refactor/indexOf 대신 contains 사용

Commit:
refactor/Game, Hint class

Jaeyoung22 commented on src/main/java/baseball/Application.java:
```diff
@@ -1,7 +1,29 @@
+
 public class Application {
+
+    private static final String PROMPT_CONTINUE = "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.";
```
:따봉

Jaeyoung22 commented on src/main/java/baseball/Application.java:
```diff
@@ -1,7 +1,29 @@
+        System.out.println(PROMPT_CONTINUE);
+        String continueInput = Console.readLine();
+
+        if (!continueInput.matches("[12]")) {
```
정규식으로 하는 방법도 있네요! 배워갑니다.
혹시 정규식을 사용함으로써 생기는 단점은 없을까요?

seank200 replied:
[길고 복잡한 정규식 🤦](https://stackoverflow.com/questions/2245282/what-is-the-longest-regular-expression-you-have-seen)의 경우 가독성 문제가 발생한다고 들었습니다.

Jaeyoung22 commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,62 @@
+    public void play() {
+        boolean isCorrect = false;
+
+        while (!isCorrect) {
```
조건문을 부정으로 시작하면 가독성이 떨어진다는 의견이 많은 것 같아요. 변수명을 `isOngoing` 등으로 두어서 조건문이 긍정으로 시작하도록 변경하는 건 어떨까요?

seank200 replied:
좋은 지적이네요 수정하겠습니다

Jaeyoung22 commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,62 @@
+        while (answer.size() < 3) {
+            int n = Randoms.pickNumberInRange(1, 9);
+            while (answer.contains(n)) {
+                n = Randoms.pickNumberInRange(1, 9);
```
PR에서 질문하셨던 것에 대한 답변 드리자면,
지역 변수에 대한 재할당은 예기치못한 side-effect를 불러올 수 있을 가능성이 있어서 IDE에서 warning 하는 것으로 알고 있습니다! 해당 로직에서도 이 while문 밖의 변수 할당이 필요해보이지는 않네요?

seank200 replied:
두번째 while문이 불필요하여 리팩토링하면서 지역변수 재할당도 삭제하였습니다

Jaeyoung22 commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,62 @@
+        }
+
+        return answer
+                .stream()
```
객체지향 생활체조를 지키려는 모습 정말 좋습니다..! 근데 저는 .stream()은 옆에 두는 편이긴 합니다ㅋㅋㅋ 한줄에 . 하나는 허용이라서 ㅎㅎ

Jaeyoung22 commented on src/main/java/baseball/Hint.java:
```diff
@@ -0,0 +1,62 @@
+            char guessDigit = guess.charAt(i);
+
+            if (guessDigit == answerDigit) {
+                strike++;
```
어디선가 느껴지는 C의 향기

seank200 replied:
바로 들켰네요 ㄷㄷ

Jaeyoung22 commented on src/main/java/baseball/Hint.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public String toString() {
```
toString() 오버라이딩 아주 자바스럽고 좋은데요?

저는 view를 따로 빼서 객체 자체를 던져 주고 view에서 필요한 정보만 꺼내 쓰도록 했는데 오버라이딩 하는 방법도 좋아보이네요

Jaeyoung22 reviewed (commented):
영우님, 코드 잘 읽었습니다! 깔끔하게 잘 구현해주신 것 같아요. 

다만 main() 메서드가 조금 비대해진 것 같은데, 책임과 역할을 살짝만 더 분리해보는 것도 좋을 것 같아요

1주차 고생하셨습니다~!

Commit:
refactor/reflect PR feedback

- Simpler main Application
- Remove unnecessary local variable reassignments
- styling: one method per line in method chain

asdftyui commented:
영우님, 코드는 재밌게 잘 읽었습니다!

Game 클래스에 너무 많은 기능이 포함되어 있어서 코드를 파악하는데 시간이 조금 오래 걸리는 것 같아요

다음 미션 때도 좋은 코드 기대하겠습니다
